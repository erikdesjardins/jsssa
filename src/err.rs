use std::fmt::{self, Debug, Display};

pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;

/// Delegates `Debug` to `Display`, since `Termination` always uses `Debug` to print errors
pub struct NiceError(Error);

impl Debug for NiceError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl<T: Into<Error>> From<T> for NiceError {
    fn from(display: T) -> Self {
        NiceError(display.into())
    }
}
