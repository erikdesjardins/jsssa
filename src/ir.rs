use swc_atoms::JsWord;

pub use self::print::print;
pub use self::ref_::{Mutable, Ref, RefType, SSA};

mod print;
mod ref_;
pub mod scope;
pub mod visit;

#[derive(Debug)]
pub struct Block(pub Vec<Stmt>);

#[derive(Debug)]
pub enum Stmt {
    Expr {
        target: Ref<SSA>,
        expr: Expr,
    },
    DeclareMutable {
        target: Ref<Mutable>,
        val: Ref<SSA>,
    },
    WriteMutable {
        target: Ref<Mutable>,
        val: Ref<SSA>,
    },
    WriteGlobal {
        target: JsWord,
        val: Ref<SSA>,
    },
    WriteMember {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
        val: Ref<SSA>,
    },
    Return {
        val: Ref<SSA>,
    },
    Throw {
        val: Ref<SSA>,
    },
    Break,
    Continue,
    Debugger,
    Block {
        body: Block,
    },
    Loop {
        body: Block,
    },
    ForEach {
        kind: ForKind,
        init: Ref<SSA>,
        body: Block,
    },
    IfElse {
        cond: Ref<SSA>,
        cons: Block,
        alt: Block,
    },
    Try {
        body: Block,
        catch: Block,
        finally: Box<Block>,
    },
}

#[derive(Debug)]
pub enum Expr {
    Bool {
        value: bool,
    },
    Number {
        value: f64,
    },
    String {
        value: JsWord,
        has_escape: bool,
    },
    Null,
    Undefined,
    This,
    Read {
        source: Ref<SSA>,
    },
    ReadMutable {
        source: Ref<Mutable>,
    },
    ReadGlobal {
        source: JsWord,
    },
    ReadMember {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
    },
    Array {
        elems: Vec<Option<(EleKind, Ref<SSA>)>>,
    },
    Object {
        props: Vec<(PropKind, Ref<SSA>, Ref<SSA>)>,
    },
    RegExp {
        regex: JsWord,
        has_escape: bool,
        flags: JsWord,
    },
    Unary {
        op: UnaryOp,
        val: Ref<SSA>,
    },
    Binary {
        op: BinaryOp,
        left: Ref<SSA>,
        right: Ref<SSA>,
    },
    Delete {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
    },
    Yield {
        kind: YieldKind,
        val: Ref<SSA>,
    },
    Await {
        val: Ref<SSA>,
    },
    Call {
        kind: CallKind,
        func: Ref<SSA>,
        args: Vec<(EleKind, Ref<SSA>)>,
    },
    Function {
        kind: FnKind,
        body: Block,
    },
    CurrentFunction,
    Argument {
        index: usize,
    },
}

#[derive(Debug)]
pub enum UnaryOp {
    Plus,
    Minus,
    Not,
    Tilde,
    Typeof,
    Void,
}

#[derive(Debug)]
pub enum BinaryOp {
    EqEq,
    NotEq,
    StrictEq,
    NotStrictEq,
    Lt,
    LtEq,
    Gt,
    GtEq,
    ShiftLeft,
    ShiftRight,
    ShiftRightZero,
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    BitOr,
    BitXor,
    BitAnd,
    Exp,
    In,
    Instanceof,
}

#[derive(Debug)]
pub enum ForKind {
    In,
    Of,
}

#[derive(Debug)]
pub enum EleKind {
    Single,
    Spread,
}

#[derive(Debug)]
pub enum PropKind {
    Simple,
    Get,
    Set,
}

#[derive(Debug)]
pub enum CallKind {
    Call,
    New,
}

#[derive(Debug)]
pub enum FnKind {
    Func { is_async: bool, is_generator: bool },
    Arrow { is_async: bool },
}

#[derive(Debug)]
pub enum YieldKind {
    Single,
    Delegate,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stmt_is_512_bits() {
        assert_eq!(std::mem::size_of::<Stmt>(), 64);
    }
}
