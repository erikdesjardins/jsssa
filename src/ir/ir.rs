use std::fmt::{self, Debug};
use std::marker::PhantomData;
use std::rc::Rc;
use std::sync::atomic::{AtomicUsize, Ordering};

// references can either be SSA or mutable (needed to model closures)
pub trait RefType {}
impl RefType for SSA {}
impl RefType for Mutable {}
#[derive(Clone, Debug)]
pub enum SSA {}
#[derive(Clone, Debug)]
pub enum Mutable {}

#[derive(Clone)]
pub enum Ref<T: RefType> {
    Dead,
    Live(LiveRef<T>),
}

impl<T: RefType> Debug for Ref<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        let mut f = f.debug_tuple("Ref");
        match self {
            Ref::Dead => {
                f.field(&"Dead");
            }
            Ref::Live(LiveRef(inner)) => {
                f.field(&inner.id);
                f.field(&inner.name_hint);
            }
        }
        f.finish()
    }
}

impl<T: RefType> Ref<T> {
    pub fn new(name_hint: String) -> Self {
        Ref::Live(LiveRef::new(name_hint))
    }

    pub fn dead() -> Self {
        Ref::Dead
    }
}

#[derive(Clone)]
struct LiveRef<T: RefType>(Rc<LiveRefInner<T>>);

struct LiveRefInner<T: RefType> {
    id: usize,
    name_hint: String,
    _t: PhantomData<T>,
}

impl<T: RefType> PartialEq for LiveRef<T> {
    fn eq(&self, other: &Self) -> bool {
        // compare only by id, which is unique by construction
        self.0.id == self.0.id
    }
}

impl<T: RefType> LiveRef<T> {
    fn new(name_hint: String) -> Self {
        static ID_COUNTER: AtomicUsize = AtomicUsize::new(1);

        LiveRef(Rc::new(LiveRefInner {
            id: ID_COUNTER.fetch_add(1, Ordering::Relaxed),
            name_hint,
            _t: PhantomData,
        }))
    }

    pub fn name_hint(&self) -> &str {
        &self.0.name_hint
    }

    pub fn might_be_used(&self) -> bool {
        Rc::strong_count(&self.0) > 1
    }
}

#[derive(Debug)]
pub struct Block {
    pub directives: Vec<String>,
    pub children: Vec<Stmt>,
}

impl Block {
    pub fn empty() -> Self {
        Block {
            directives: vec![],
            children: vec![],
        }
    }
}

#[derive(Debug)]
pub enum Stmt {
    Expr(Expr),
    Assign(Ref<SSA>, Expr),
    AssignBinding(Ref<Mutable>, Ref<SSA>),
    AssignGlobal(String, Ref<SSA>),
    Return(Ref<SSA>),
    Throw(Ref<SSA>),
    Debugger,
    Block(Box<Block>),
    IfElse(Ref<SSA>, Box<Block>, Box<Block>),
    Try(
        Box<Block>,
        Option<(Ref<Mutable>, Box<Block>)>,
        Option<Box<Block>>,
    ),
}

#[derive(Debug)]
pub enum Expr {
    Bool(bool),
    Number(f64),
    String(String),
    Null,
    Undefined,
    This,
    Read(Ref<SSA>),
    ReadBinding(Ref<Mutable>),
    ReadGlobal(String),
    Array(Vec<Option<(EleKind, Ref<SSA>)>>),
    Object(Vec<(PropKind, Ref<SSA>, Ref<SSA>)>),
    RegExp(String, String),
    Unary(UnaryOp, Ref<SSA>),
    Binary(BinaryOp, Ref<SSA>, Ref<SSA>),
    Call(CallKind, Ref<SSA>, Vec<Ref<SSA>>),
    Function(FnKind, Option<String>, Vec<Ref<Mutable>>, Box<Block>),
    Yield(YieldKind, Ref<SSA>),
    Await(Ref<SSA>),
}

#[derive(Debug)]
pub enum UnaryOp {
    Plus,
    Minus,
    Not,
    Tilde,
    Typeof,
    Void,
    Delete,
}

#[derive(Debug)]
pub enum BinaryOp {
    Eq,
    NotEq,
    StrictEq,
    NotStrictEq,
    LT,
    LTE,
    GT,
    GTE,
    ShiftLeft,
    ShiftRight,
    ShiftRightZero,
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    BitOr,
    BitXor,
    BitAnd,
    In,
    Instanceof,
}

#[derive(Debug)]
pub enum EleKind {
    Single,
    Spread,
}

#[derive(Debug)]
pub enum PropKind {
    Simple,
    Get,
    Set,
}

#[derive(Debug)]
pub enum CallKind {
    Call,
    New,
}

#[derive(Debug)]
pub enum FnKind {
    Func { async: bool, gen: bool },
    Arrow { async: bool },
}

#[derive(Debug)]
pub enum YieldKind {
    Single,
    Delegate,
}
