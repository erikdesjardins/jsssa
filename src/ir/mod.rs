mod types;
mod ref_;

pub use self::types::*;
pub use self::ref_::*;
