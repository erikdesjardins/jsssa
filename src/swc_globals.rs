/// A token proving that SWC's globals are initialized.
pub struct Initialized(());

/// Initializes the globals used by SWC and provides a token as proof.
///
/// A higher-rank lifetime ensures the token cannot escape the closure.
pub fn with<R>(f: impl FnOnce(&'_ Initialized) -> R) -> R {
    swc_common::GLOBALS.set(&swc_common::Globals::new(), || f(&Initialized(())))
}
