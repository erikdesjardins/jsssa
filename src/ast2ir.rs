use std::collections::HashMap;

use ast;
use ir;
use util::*;

type ScopeMap = Map<String, ir::Ref<ir::Mutable>>;

pub fn convert(ast: ast::File) -> ir::Block {
    let ast::File {
        program:
            ast::Program {
                body,
                directives,
                source_type: _,
            },
    } = ast;

    convert_block(body, directives, &HashMap::default())
}

fn convert_block(
    body: Vec<ast::Statement>,
    directives: Vec<ast::Directive>,
    parent_scopes: &ScopeMap,
) -> ir::Block {
    let mut scope = parent_scopes.clone();

    let children = body.into_iter()
        .flat_map(|stmt| convert_statement(stmt, &mut scope))
        .collect();

    ir::Block {
        directives: directives.into_iter().map(|d| d.value.value).collect(),
        children,
    }
}

fn convert_statement(stmt: ast::Statement, scope: &mut ScopeMap) -> Vec<ir::Stmt> {
    use ast::Statement::*;

    match stmt {
        ExpressionStatement(ast::ExpressionStatement { expression }) => {
            let (mut exprs, last_expr) = convert_expression(expression, scope);
            exprs.push(ir::Stmt::Expr(last_expr));
            exprs
        },
        BlockStatement(ast::BlockStatement { body, directives }) =>
            vec![ir::Stmt::Block(box convert_block(body, directives, scope))],
        EmptyStatement(ast::EmptyStatement {}) =>
            vec![],
        DebuggerStatement(ast::DebuggerStatement {}) =>
            vec![ir::Stmt::Debugger],
        WithStatement(_) =>
            unimplemented!("with() statement not supported"),
        ReturnStatement(ast::ReturnStatement { argument }) => {
            let ref_ = ir::Ref::new("return_".to_string());
            match argument {
                Some(argument) => {
                    let (mut exprs, return_value) = convert_expression(argument, scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    exprs
                },
                None => {
                    vec![
                        ir::Stmt::Assign(ref_.clone(), ir::Expr::Undefined),
                        ir::Stmt::Return(ref_),
                    ]
                }
            }
        },
        LabeledStatement(_) |
        BreakStatement(_) |
        ContinueStatement(_) =>
            unimplemented!("labels not yet supported"),
        IfStatement(ast::IfStatement { test, consequent, alternate }) => {
            let ref_ = ir::Ref::new("if_".to_string());
            let (mut exprs, test_value) = convert_expression(test, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), test_value));
            exprs.push(ir::Stmt::IfElse(
                ref_,
                {
                    let children = convert_statement(*consequent, scope);
                    box ir::Block { directives: vec![], children }
                },
                match alternate {
                    Some(alternate) => {
                        let children = convert_statement(*alternate, scope);
                        box ir::Block { directives: vec![], children }
                    },
                    None => box ir::Block::empty()
                },
            ));
            exprs
        },
        SwitchStatement(_) =>
            // remember, switch cases aren't evaluated (!) until they're checked for equality
            unimplemented!("switch statements not yet supported"),
        ThrowStatement(ast::ThrowStatement { argument }) => {
            let ref_ = ir::Ref::new("throw_".to_string());
            let (mut exprs, throw_value) = convert_expression(argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), throw_value));
            exprs.push(ir::Stmt::Throw(ref_));
            exprs
        },
        TryStatement(ast::TryStatement { block, handler, finalizer }) => {
            let try = ir::Stmt::Try(
                {
                    let ast::BlockStatement { body, directives } = block;
                    box convert_block(body, directives, scope)
                },
                handler.map(|ast::CatchClause { param, body }| {
                    let ast::BlockStatement { body, directives } = body;
                    let mut catch_scope = scope.clone();
                    let ref_ = convert_pattern(param, &mut catch_scope);
                    (ref_, box convert_block(body, directives, &catch_scope))
                }),
                finalizer.map(|finalizer| {
                    let ast::BlockStatement { body, directives } = finalizer;
                    box convert_block(body, directives, scope)
                }),
            );
            vec![try]
        },
        _ => unimplemented!(),
    }
}

fn convert_expression(expr: ast::Expression, scope: &ScopeMap) -> (Vec<ir::Stmt>, ir::Expr) {
    use ast::Expression::*;

    match expr {
        Identifier(ast::Identifier { name }) => {
            let expr = match scope.get(&name) {
                Some(ref_) => ir::Expr::ReadBinding(ref_.clone()),
                None => ir::Expr::ReadGlobal(name),
            };
            (vec![], expr)
        }
        RegExpLiteral(ast::RegExpLiteral { pattern, flags }) => {
            (vec![], ir::Expr::RegExp(pattern, flags))
        }
        NullLiteral(ast::NullLiteral {}) => (vec![], ir::Expr::Null),
        StringLiteral(ast::StringLiteral { value }) => (vec![], ir::Expr::String(value)),
        BooleanLiteral(ast::BooleanLiteral { value }) => (vec![], ir::Expr::Bool(value)),
        NumericLiteral(ast::NumericLiteral { value }) => (vec![], ir::Expr::Number(value)),
        ThisExpression(ast::ThisExpression {}) => (vec![], ir::Expr::This),
        ArrowFunctionExpression(ast::ArrowFunctionExpression {
            params,
            body,
            async,
        }) => {
            use ast::BlockStmtOrExpr::*;

            let mut fn_scope = scope.clone();
            let refs = params
                .into_iter()
                .map(|param| convert_pattern(param, &mut fn_scope))
                .collect();
            let body = match *body {
                BlockStatement(block) => {
                    let ast::BlockStatement { body, directives } = block;
                    convert_block(body, directives, &fn_scope)
                }
                Expression(expr) => {
                    let ref_ = ir::Ref::new("arrow_".to_string());
                    let (mut exprs, return_value) = convert_expression(expr, &fn_scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    ir::Block {
                        directives: vec![],
                        children: exprs,
                    }
                }
            };
            let func = ir::Expr::Function(ir::FnKind::Arrow { async }, None, refs, box body);
            (vec![], func)
        }
        YieldExpression(ast::YieldExpression { argument, delegate }) => {
            let ref_ = ir::Ref::new("yield_".to_string());
            let (mut exprs, yield_value) = match argument {
                Some(argument) => convert_expression(*argument, scope),
                None => (vec![], ir::Expr::Undefined),
            };
            exprs.push(ir::Stmt::Assign(ref_.clone(), yield_value));
            let kind = if delegate {
                ir::YieldKind::Delegate
            } else {
                ir::YieldKind::Single
            };
            (exprs, ir::Expr::Yield(kind, ref_))
        }
        AwaitExpression(ast::AwaitExpression { argument }) => {
            let ref_ = ir::Ref::new("await_".to_string());
            let (mut exprs, await_value) = convert_expression(*argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), await_value));
            (exprs, ir::Expr::Await(ref_))
        }
        _ => unimplemented!(),
    }
}

fn convert_pattern(pat: ast::Pattern, scope: &mut ScopeMap) -> ir::Ref<ir::Mutable> {
    use ast::Pattern::*;

    match pat {
        Identifier(ast::Identifier { name }) => {
            let ref_ = ir::Ref::new(name.clone());
            scope.insert(name, ref_.clone());
            ref_
        }
        MemberExpression(_) | ObjectPattern(_) | ArrayPattern(_) | RestElement(_)
        | AssignmentPattern(_) => unimplemented!("complex patterns not yet supported"),
    }
}
