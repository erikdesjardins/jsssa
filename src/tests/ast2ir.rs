use crate::ast2ir;
use crate::err::NiceError;
use crate::ir;
use crate::parse;
use crate::swc_globals;

macro_rules! case {
    ( $name:ident, $string:expr, @ $expected:literal ) => {
        #[test]
        fn $name() -> Result<(), NiceError> {
            swc_globals::with(|g| {
                let (ast, _) = parse::parse(g, $string)?;
                let ir = ast2ir::convert(g, ast);
                let ppr = ir::print(g, &ir);
                insta::assert_snapshot!(ppr, @ $expected);
                Ok(())
            })
        }
    };
}

case!(
    basic,
    r#"
    function f(x) {
        while (true);
        x = y.bar;
        z.foo = x ? true : 'hi';
        return +[1 || x, { x }, f + 1, ++g];
    }
    f(1), true;
"#,
@r###"
_ini = <void>
f <= _ini
_fun = <function>:
    f$1 = <current function>
    f$2 <= f$1
    x = <argument 0>
    x$1 <= x
    <loop>:
        _whl = true
        <if> _whl:
            <empty>
        <else>:
            <break>
    _obj = <global y>
    _prp = "bar"
    _val = _obj[_prp]
    x$1 <- _val
    <dead> = _val
    _obj$1 = <global z>
    _prp$1 = "foo"
    _tst = *x$1
    _udf = <void>
    _val$1 <= _udf
    <if> _tst:
        _cns = true
        _val$1 <- _cns
    <else>:
        _alt = "hi"
        _val$1 <- _alt
    _val$2 = *_val$1
    _obj$1[_prp$1] <- _val$2
    <dead> = _val$2
    _prd = 1
    _log <= _prd
    <if> _prd:
        <empty>
    <else>:
        _cns = *x$1
        _log <- _cns
    _ele = *_log
    _key = "x"
    _val$3 = *x$1
    _ele$1 = { [_key]: _val$3 }
    _lhs = *f$2
    _rhs = 1
    _ele$2 = _lhs + _rhs
    _rdr = <global g>
    _one = 1
    _wri = _rdr + _one
    <global g> <- _wri
    _ele$3 = _wri
    _una = [_ele, _ele$1, _ele$2, _ele$3]
    _ret = + _una
    <return> _ret
f <- _fun
_fun$1 = *f
_arg = 1
<dead> = _fun$1(_arg)
<dead> = true
"###);

case!(
    object_props,
    r#"
    o.x;
    o[x];
    o.x = 1;
    o[x] = 1;
    delete o.x;
    delete o[x];
"#,
@r###"
_obj = <global o>
_prp = "x"
<dead> = _obj[_prp]
_obj$1 = <global o>
_prp$1 = <global x>
<dead> = _obj$1[_prp$1]
_obj$2 = <global o>
_prp$2 = "x"
_val = 1
_obj$2[_prp$2] <- _val
<dead> = _val
_obj$3 = <global o>
_prp$3 = <global x>
_val$1 = 1
_obj$3[_prp$3] <- _val$1
<dead> = _val$1
_obj$4 = <global o>
_prp$4 = "x"
<dead> = <delete> _obj$4[_prp$4]
_obj$5 = <global o>
_prp$5 = <global x>
<dead> = <delete> _obj$5[_prp$5]
"###);

case!(
    objects,
    r#"
    ({ x, y: 1, [z]: 2 })
"#,
@r###"
_key = "x"
_val = <global x>
_key$1 = "y"
_val$1 = 1
_key$2 = <global z>
_val$2 = 2
<dead> = { [_key]: _val, [_key$1]: _val$1, [_key$2]: _val$2 }
"###);

case!(
    deep_scopes,
    r#"
    var x = 1;
    (function() {
        (function() {
            x = 2;
        });
    });
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = 1
x <- _ini$1
<dead> = <function>:
    <dead> = <function>:
        _val = 2
        x <- _val
        <dead> = _val
"###);

case!(
    var_reassignment_1,
    r#"
    var x = 1;
    const y = 1;
    {
        var x = 2;
        const y = 2;
    }
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = 1
x <- _ini$1
_ini$2 = 1
y <= _ini$2
_ini$3 = 2
x <- _ini$3
_ini$4 = 2
y$1 <= _ini$4
"###);

case!(
    var_reassignment_2,
    r#"
    var x = 1;
    const y = 1;
    (function() {
        var x = 2;
        const y = 2;
    });
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = 1
x <- _ini$1
_ini$2 = 1
y <= _ini$2
<dead> = <function>:
    _ini$3 = <void>
    x$1 <= _ini$3
    _ini$4 = 2
    x$1 <- _ini$4
    _ini$5 = 2
    y$1 <= _ini$5
"###);

case!(
    var_scoping_1,
    r#"
    var x = 1;
    {
        var x = 2;
    }
    x = 3;
    var x = 4;
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = 1
x <- _ini$1
_ini$2 = 2
x <- _ini$2
_val = 3
x <- _val
<dead> = _val
_ini$3 = 4
x <- _ini$3
"###);

case!(
    var_scoping_2,
    r#"
    var x = 1;
    {
        let x = 2;
        x = 3;
    }
    x = 4;
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = 1
x <- _ini$1
_ini$2 = 2
x$1 <= _ini$2
_val = 3
x$1 <- _val
<dead> = _val
_val$1 = 4
x <- _val$1
<dead> = _val$1
"###);

case!(
    var_hoisting_1,
    r#"
    x = 1;
    {
        var x = 2;
    }
    x = 3;
"#,
@r###"
_ini = <void>
x <= _ini
_val = 1
x <- _val
<dead> = _val
_ini$1 = 2
x <- _ini$1
_val$1 = 3
x <- _val$1
<dead> = _val$1
"###);

case!(
    var_hoisting_2,
    r#"
    x = 1;
    {
        for (var x of 2);
    }
    x = 3;
"#,
@r###"
_ini = <void>
x <= _ini
_val = 1
x <- _val
<dead> = _val
_rhs = 2
<foreach of> _rhs:
    _for = <argument 0>
    x <- _for
_val$1 = 3
x <- _val$1
<dead> = _val$1
"###);

case!(
    for_in_no_var,
    r#"
    var x;
    for (x in 1);
    for (y in 1);
"#,
@r###"
_ini = <void>
x <= _ini
_ini$1 = <void>
x <- _ini$1
_rhs = 1
<foreach in> _rhs:
    _for = <argument 0>
    x <- _for
_rhs$1 = 1
<foreach in> _rhs$1:
    _for = <argument 0>
    <global y> <- _for
"###);

case!(
    conditional,
    r#"
    (cond ? true_ : false_);
"#,
@r###"
_tst = <global cond>
_udf = <void>
_val <= _udf
<if> _tst:
    _cns = <global true_>
    _val <- _cns
<else>:
    _alt = <global false_>
    _val <- _alt
<dead> = *_val
"###);

case!(
    logical_op,
    r#"
    (foo || or_else);
    (bar && and_then);
"#,
@r###"
_prd = <global foo>
_log <= _prd
<if> _prd:
    <empty>
<else>:
    _cns = <global or_else>
    _log <- _cns
<dead> = *_log
_prd$1 = <global bar>
_log$1 <= _prd$1
<if> _prd$1:
    _cns = <global and_then>
    _log$1 <- _cns
<else>:
    <empty>
<dead> = *_log$1
"###);

case!(
    assign_to_expr,
    r#"
    e |= 0;
    foo().x |= 1;
"#,
@r###"
_lhs = <global e>
_rhs = 0
_val = _lhs | _rhs
<global e> <- _val
<dead> = _val
_fun = <global foo>
_obj = _fun()
_prp = "x"
_lhs$1 = _obj[_prp]
_rhs$1 = 1
_val$1 = _lhs$1 | _rhs$1
_obj[_prp] <- _val$1
<dead> = _val$1
"###);

case!(
    labels,
    r#"
    outer: for (;;) {
        inner: for (;;) {
            if (foo) continue inner;
            if (bar) break outer;
        }
    }
"#,
@r###"
<label outer>:
    <loop>:
        <label inner>:
            <loop>:
                _iff = <global foo>
                <if> _iff:
                    <continue inner>
                <else>:
                    <empty>
                _iff$1 = <global bar>
                <if> _iff$1:
                    <break outer>
                <else>:
                    <empty>
"###);

case!(
    referencing_outer_scope_declared_later,
    r#"
    g = function() {
        x;
        y;
        z;
    };
    var x = 0;
    let y = 1;
    const z = 2;
"#,
@r###"
_ini = <void>
x <= _ini
_val = <function>:
    <dead> = *x
    <dead> = *y
    <dead> = *z
<global g> <- _val
<dead> = _val
_ini$1 = 0
x <- _ini$1
_ini$2 = 1
y <= _ini$2
_ini$3 = 2
z <= _ini$3
"###);

case!(
    referencing_outer_scope_declared_later2,
    r#"
    g = function() {
        x;
        y; // global
        z; // global
    };
    {
        var x = 0;
        let y = 1;
        const z = 2;
    }
"#,
@r###"
_ini = <void>
x <= _ini
_val = <function>:
    <dead> = *x
    <dead> = <global y>
    <dead> = <global z>
<global g> <- _val
<dead> = _val
_ini$1 = 0
x <- _ini$1
_ini$2 = 1
y <= _ini$2
_ini$3 = 2
z <= _ini$3
"###);

case!(
    mutually_recursive_fns,
    r#"
    function a() { b(); }
    function b() { c(); }
    function c() { a(); }
    g1 = a;
    g2 = b;
    g3 = c;
"#,
@r###"
_ini = <void>
a <= _ini
_ini$1 = <void>
b <= _ini$1
_ini$2 = <void>
c <= _ini$2
_fun = <function>:
    a$1 = <current function>
    a$2 <= a$1
    _fun$3 = *b
    <dead> = _fun$3()
a <- _fun
_fun$1 = <function>:
    b$1 = <current function>
    b$2 <= b$1
    _fun$3 = *c
    <dead> = _fun$3()
b <- _fun$1
_fun$2 = <function>:
    c$1 = <current function>
    c$2 <= c$1
    _fun$3 = *a
    <dead> = _fun$3()
c <- _fun$2
_val = *a
<global g1> <- _val
<dead> = _val
_val$1 = *b
<global g2> <- _val$1
<dead> = _val$1
_val$2 = *c
<global g3> <- _val$2
<dead> = _val$2
"###);

case!(
    fn_hoisting_toplevel,
    r#"
    foo();
    function foo() { foo_; }

    (function() {
        bar();
        function bar() { bar_; }
    })();
"#,
@r###"
_ini = <void>
foo <= _ini
_fun = <function>:
    foo$1 = <current function>
    foo$2 <= foo$1
    <dead> = <global foo_>
foo <- _fun
_fun$1 = *foo
<dead> = _fun$1()
_fun$2 = <function>:
    _ini$1 = <void>
    bar <= _ini$1
    _fun$3 = <function>:
        bar$1 = <current function>
        bar$2 <= bar$1
        <dead> = <global bar_>
    bar <- _fun$3
    _fun$4 = *bar
    <dead> = _fun$4()
<dead> = _fun$2()
"###);

case!(
    fn_hoisting_blocks,
    r#"
    if (x) {
        foo();
        function foo() { foo_; }
    }
    foo();
"#,
@r###"
_ini = <void>
foo <= _ini
_iff = <global x>
<if> _iff:
    _fun$1 = *foo
    <dead> = _fun$1()
    _fun$2 = <function>:
        foo$1 = <current function>
        foo$2 <= foo$1
        <dead> = <global foo_>
    foo <- _fun$2
<else>:
    <empty>
_fun = *foo
<dead> = _fun()
"###);

case!(
    switch,
    r#"
    switch (x) {
        case 1:
            one;
            break;
        case "foo":
        case bar:
            two;
        default:
            def;
    }
"#,
@r###"
_swi = <global x>
_tst = 1
_tst$1 = "foo"
_tst$2 = <global bar>
<switch> _swi:
    <case> _tst:
    <dead> = <global one>
    <break>
    <case> _tst$1:
    <case> _tst$2:
    <dead> = <global two>
    <default>:
    <dead> = <global def>
"###);

case!(
    switch_scoping_forwards,
    r#"
    switch (x) {
        case 1:
            var v = 2;
            let l = 3;
        default:
            g1 = v;
            g2 = l;
    }
"#,
@r###"
_ini = <void>
v <= _ini
_swi = <global x>
_tst = 1
<switch> _swi:
    <case> _tst:
    _ini$1 = 2
    v <- _ini$1
    _ini$2 = 3
    l <= _ini$2
    <default>:
    _val = *v
    <global g1> <- _val
    <dead> = _val
    _val$1 = *l
    <global g2> <- _val$1
    <dead> = _val$1
"###);

case!(
    switch_scoping_backwards,
    r#"
    switch (x) {
        case 1:
            g1 = v;
            g2 = l;
            break;
        default:
            var v = 2;
            let l = 3;
    }
"#,
@r###"
_ini = <void>
v <= _ini
_swi = <global x>
_tst = 1
<switch> _swi:
    <case> _tst:
    _val = *v
    <global g1> <- _val
    <dead> = _val
    _val$1 = *l
    <global g2> <- _val$1
    <dead> = _val$1
    <break>
    <default>:
    _ini$1 = 2
    v <- _ini$1
    _ini$2 = 3
    l <= _ini$2
"###);

case!(
    preserves_prop_calls,
    r#"
    console.log.bind(console);
"#,
@r###"
_obj = <global console>
_prp = "log"
_obj$1 = _obj[_prp]
_prp$1 = "bind"
_arg = <global console>
<dead> = _obj$1[_prp$1](_arg)
"###);

case!(
    does_not_preserve_indirect_calls,
    r#"
    (0, console.log)();
"#,
@r###"
<dead> = 0
_obj = <global console>
_prp = "log"
_fun = _obj[_prp]
<dead> = _fun()
"###);

case!(
    does_not_preserve_new,
    r#"
    new console.log();
"#,
@r###"
_obj = <global console>
_prp = "log"
_fun = _obj[_prp]
<dead> = <new> _fun()
"###);

case!(
    arg_shadow_fn_name_decl,
    r#"
    function f(f, a) {
        f(a);
    }
    g = f;
"#,
@r###"
_ini = <void>
f <= _ini
_fun = <function>:
    f$1 = <current function>
    f$2 <= f$1
    f$3 = <argument 0>
    f$2 <- f$3
    a = <argument 1>
    a$1 <= a
    _fun$1 = *f$2
    _arg = *a$1
    <dead> = _fun$1(_arg)
f <- _fun
_val = *f
<global g> <- _val
<dead> = _val
"###);

case!(
    arg_shadow_fn_name_expr,
    r#"
    g = function f(f, a) {
        f(a);
    };
"#,
@r###"
_val = <function>:
    f = <current function>
    f$1 <= f
    f$2 = <argument 0>
    f$1 <- f$2
    a = <argument 1>
    a$1 <= a
    _fun = *f$1
    _arg = *a$1
    <dead> = _fun(_arg)
<global g> <- _val
<dead> = _val
"###);

case!(
    decr_object,
    r#"
    obj.x--;
"#,
@r###"
_obj = <global obj>
_prp = "x"
_rdr = _obj[_prp]
_one = 1
_wri = _rdr - _one
_obj[_prp] <- _wri
<dead> = _rdr
"###);

case!(
    incr_object,
    r#"
    ++obj.x;
"#,
@r###"
_obj = <global obj>
_prp = "x"
_rdr = _obj[_prp]
_one = 1
_wri = _rdr + _one
_obj[_prp] <- _wri
<dead> = _wri
"###);
