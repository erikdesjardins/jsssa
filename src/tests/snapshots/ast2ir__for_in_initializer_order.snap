---
created: "2019-03-30T21:40:02.321785900Z"
creator: insta@0.7.1
source: src/tests/ast2ir.rs
expression: "\n    for (var x in b);\n"
---
_ini = <void>
x :- _ini
_rhs = <global b>
<foreach in> _rhs:
    _for = <argument 0>
    x <- _for
